datumaro/plugins/open_images_format.py

```diff
@@ -398,12 +398,22 @@ def __init__(self, root_dir):
 
         os.makedirs(self._annotations_dir, exist_ok=True)
 
+    @contextlib.contextmanager
     def open(self, file_name, newline=None):
         self._written_annotations.add(file_name)
-        f = open(
-            osp.join(self._annotations_dir, file_name),
-            'w', encoding='utf-8', newline=newline)
-        return f
+
+        file_path = osp.join(self._annotations_dir, file_name)
+
+        # Write to a temporary file first, to avoid data loss if we're patching
+        # an existing dataset and the process is interrupted.
+        temp_file_path = file_path + '.tmp'
```

Comment:
I wouldn't say it is a problem, but using https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp would be more pragmatic. Consider using id with `datumaro.utils.error_cleanup`.

Reply:
> I wouldn't say it is a problem, but using https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp would be more pragmatic.

I've considered it, but `mkstemp` (and its various derivatives) has one quality that makes it unsuitable for this use case: it always creates the temporary file with `600` permissions, which are not necessarily the same permissions as one would get by creating the file normally. And correcting the permissions after the fact is surprisingly involved (the default permissions depend on the process umask, and you can't actually fetch the current umask without changing it, and changing it would affect other threads...).

Furthermore, the only benefit of using it would be avoiding filename conflicts, but the only way such conflicts would occur would be if either:

a. two datasets were being written to the same directory at the same time, which I'm pretty sure would break horribly anyway, or
b. a `.tmp` file already existed in the `annotations` directory and contained valuable data, which seems unlikely (there are certainly no `.tmp` files in the original Open Images dataset).

> Consider using it with datumaro.utils.error_cleanup.

That doesn't seem to exist. Perhaps it's called something else?

Reply:
> That doesn't seem to exist. Perhaps it's called something else?

I meant this: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/util/__init__.py#L192

An example: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/cli/contexts/project/__init__.py#L551